bulkclient/src/main/java/au/csiro/pathling/export/BulkExportClient.java

```diff
@@ -0,0 +1,389 @@
+
+  /**
+   * The maximum number of concurrent downloads to perform.
+   */
+  @Builder.Default
+  @Min(1)
+  int maxConcurrentDownloads = 10;
+
+  /**
+   * The factory to use to create the {@link FileStore} to write the output files to.
+   */
+  @Nonnull
+  @Builder.Default
+  FileStoreFactory fileStoreFactory = FileStoreFactory.getLocal();
+
+  /**
+   * The configuration for the HTTP client.
+   */
+  @Nonnull
+  @Builder.Default
+  HttpClientConfiguration httpClientConfig = HttpClientConfiguration.builder().build();
+
+  /**
+   * The configuration for the async operations.
+   */
+  @Nonnull
+  @Builder.Default
+  AsyncConfig asyncConfig = AsyncConfig.builder().build();
+
+
+  /**
+   * The configuration for the authentication.
+   */
+  @Nonnull
+  @Builder.Default
+  AuthConfiguration authConfig = AuthConfiguration.builder().build();
+
+  /**
+   * A builder for the {@link BulkExportClient}.
+   */
+  public static class BulkExportClientBuilder {
+    // empty placeholder to for javadoc to recognize the builder
+  }
+
+  /**
+   * Create a builder for a system-level export.
+   *
+   * @return the builder configured for a system-level export
+   */
+  @Nonnull
+  public static BulkExportClientBuilder systemBuilder() {
+    return BulkExportClient.builder().withLevel(new SystemLevel());
+  }
+
+  /**
+   * Create a builder for a patient-level export.
+   *
+   * @return the builder configured for a patient-level export
+   */
+  @Nonnull
+  public static BulkExportClientBuilder patientBuilder() {
+    return BulkExportClient.builder().withLevel(new PatientLevel());
+  }
+
+  /**
+   * Create a builder for a group-level export.
+   *
+   * @param groupId the group ID to export
+   * @return the builder configured for a group-level export
+   */
+  @Nonnull
+  public static BulkExportClientBuilder groupBuilder(@Nonnull final String groupId) {
+    return BulkExportClient.builder().withLevel(new GroupLevel(groupId));
+  }
+
+  /**
+   * Export data from the FHIR server.
+   *
+   * @return the result of the export
+   * @throws BulkExportException if the export fails
+   */
+  public BulkExportResult export() {
+    try (
+        final TokenCredentialFactory tokenAuthFactory = createTokenProvider();
+        final FileStore fileStore = createFileStore();
+        final CloseableHttpClient httpClient = createHttpClient(tokenAuthFactory);
+        final ExecutorServiceResource executorServiceResource = createExecutorServiceResource()
+    ) {
+      final BulkExportTemplate bulkExportTemplate = new BulkExportTemplate(
+          new BulkExportAsyncService(httpClient, URI.create(fhirEndpointUrl)),
+          asyncConfig);
+      final UrlDownloadTemplate downloadTemplate = new UrlDownloadTemplate(httpClient,
+          executorServiceResource.getExecutorService());
+
+      final BulkExportResult result = doExport(fileStore, bulkExportTemplate, downloadTemplate);
+      log.info("Export successful: {}", result);
```

Comment:
Should we issue a [Bulk Data Delete Request](https://build.fhir.org/ig/HL7/bulk-data/export.html#bulk-data-delete-request) once the download has successfully completed (if the server supports it)?

Reply:
The spec say that we MAY issue it to either cancel the request in progress or to cleanup.
So I guess it depends what we want to consider to be required for the first release of the bulk client.
I this this might be something to be added  subsequent releases.

Reply:
I think we should add this to mitigate the problem of exported data accumulating in the source server, which has been a problem that we have encountered with HAPI JPA.

Reply:
Implemented.